server/src/session.ts

```diff
@@ -131,6 +131,51 @@ export class Session {
     // project as dirty to force update the graph.
     project.markAsDirty();
     this.info(`Enabling Ivy language service for ${project.projectName}.`);
+
+    // Send diagnostics since we skipped this step when opening the file
+    // (because language service was disabled while waiting for ngcc).
+    // Technically we could just call project.refreshDiagnostics(), but we need
+    // to handle the special case where all open files are external templates.
+    const openFiles = this.getOpenFilesForNewlyLoadedProject(project);
```

Comment:
I think it would be useful to explain the entire strategy somewhere, maybe in the doc comment for this method. If I recall correctly: when someone opens an external template, the project service walks directories and finds all projects, for all projects we check if it is an Ng project and run ngcc, then we get back the ngcc complete response, and at this point we get trigger diagnostics for a typescript file in the project, which would cause an incremental build of the compiler, in the process discovering all external templates and attaching them to the project. (More to make sure I remember this :) )

Reply:
I made some minor revisions to the same doc I wrote and posted it to https://github.com/angular/vscode-ng-language-service/wiki/Project-Matching-for-External-Templates
I've also added the link to the commit and PR description.
Your description of the entire workflow is spot-on!